compiler/noirc_frontend/src/elaborator/mod.rs

```diff
@@ -779,4 +796,310 @@ impl<'context> Elaborator<'context> {
+        let impl_id = self.interner.next_trait_impl_id();
+
+        self.self_type = Some(self_type.clone());
+        self.current_trait_impl = Some(impl_id);
+
+        let mut methods = trait_impl.methods.function_ids();
+
+        self.elaborate_functions(trait_impl.methods);
+
+        let maybe_trait_id = trait_impl.trait_id;
+        if let Some(trait_id) = maybe_trait_id {
+            for func_id in &methods {
+                self.interner.set_function_trait(*func_id, self_type.clone(), trait_id);
+            }
+        }
+
+        if matches!(self_type, Type::MutableReference(_)) {
+            let span = self_type_span.unwrap_or_else(|| trait_impl.trait_path.span());
+            self.push_err(DefCollectorErrorKind::MutableReferenceInTraitImpl { span });
+        }
+
+        if let Some(trait_id) = maybe_trait_id {
+            let where_clause = trait_impl
+                .where_clause
+                .into_iter()
+                .flat_map(|item| self.resolve_trait_constraint(item))
+                .collect();
+
+            let resolved_trait_impl = Shared::new(TraitImpl {
+                ident: trait_impl.trait_path.last_segment().clone(),
+                typ: self_type.clone(),
+                trait_id,
+                trait_generics: trait_generics.clone(),
+                file: trait_impl.file_id,
+                where_clause,
+                methods,
+            });
+
+            let generics = vecmap(&self.generics, |(_, type_variable, _)| type_variable.clone());
+
+            if let Err((prev_span, prev_file)) = self.interner.add_trait_implementation(
+                self_type.clone(),
+                trait_id,
+                trait_generics,
+                impl_id,
+                generics,
+                resolved_trait_impl,
+            ) {
+                self.push_err(DefCollectorErrorKind::OverlappingImpl {
+                    typ: self_type.clone(),
+                    span: self_type_span.unwrap_or_else(|| trait_impl.trait_path.span()),
+                });
+
+                // The 'previous impl defined here' note must be a separate error currently
+                // since it may be in a different file and all errors have the same file id.
+                self.file = prev_file;
+                self.push_err(DefCollectorErrorKind::OverlappingImplNote { span: prev_span });
+                self.file = trait_impl.file_id;
+            }
+        }
+
+        self.self_type = None;
+        self.current_trait_impl = None;
+        self.generics.clear();
+    }
+
+    fn collect_impls(
+        &mut self,
+        self_type: &UnresolvedType,
+        module: LocalModuleId,
+        impls: &[(Vec<Ident>, Span, UnresolvedFunctions)],
+    ) {
+        self.local_module = module;
+
+        for (generics, span, unresolved) in impls {
+            self.file = unresolved.file_id;
+            self.declare_method_on_struct(self_type, generics, false, unresolved, *span);
```

Comment:
hmm, in that case I'd consider the function to have too many parameters

Reply:
I want to generally avoid refactorings with these PRs if possible since the longer they're open the longer we'd have to duplicate any changes going into name resolution and type checking. So I'd like to hold off on this change for now.